importer/src/path_canonicalizer.rs

```diff
@@ -302,20 +342,113 @@ fn check_machine<T: Clone>(
     state: &mut State<'_, T>,
 ) -> Result<(), String> {
     // we check the path in the context of the parent module
-    let module_location = {
-        let mut l = location.clone();
-        l.pop();
-        l
-    };
+    let module_location = location.clone().parent();
 
+    // Find all local variables.
+    let mut local_variables = HashSet::<String>::default();
+    for name in m.local_names() {
+        if !local_variables.insert(name.clone()) {
+            return Err(format!("Duplicate name `{name}` in machine `{location}`"));
+        }
+    }
     for statement in &m.statements {
-        if let MachineStatement::Submachine(_, path, _) = statement {
-            check_path(module_location.clone().join(path.clone()), state)?
+        match statement {
+            MachineStatement::Submachine(_, path, _) => {
+                check_path(module_location.clone().join(path.clone()), state)?
+            }
+            MachineStatement::Pil(_, statement) => statement
+                .expressions()
+                .try_for_each(|e| check_expression(&module_location, e, state, &local_variables))?,
+            _ => {}
         }
     }
     Ok(())
 }
 
+/// Checks an expression, checking the paths it contains.
+///
+/// Local variables are those that do not have a global path. They can be referenced by direct name only.
+///
+/// # Errors
+///
+/// This function will return an error if any of the paths does not resolve to anything
+fn check_expression<T: Clone>(
+    location: &AbsoluteSymbolPath,
+    e: &Expression<T>,
+    state: &mut State<'_, T>,
+    local_variables: &HashSet<String>,
+) -> Result<(), String> {
+    // We cannot use the visitor here because we need to change the local variables
```

Comment:
why doesn't the mut visitor work for that?

Reply:
Because we need a push/pop behavior:

When entering something that has local variables, we need to change them and then change them back when leaving that object. The mut iterator only calls once for each element, but does not care about how it is nested.

Reply:
right